File: docs/breez-sdk/snippets/swift/BreezSdkSnippets/Sources/Optimize.swift
import BreezSdkSpark

func runFullOptimization(sdk: BreezSdk) async throws {
    // ANCHOR: optimize-leaves-full
    let outcome = try await sdk.optimizeLeaves(request: OptimizeLeavesRequest(mode: .full)).outcome

    switch outcome {
    case .completed(let roundsExecuted):
        if roundsExecuted == 0 {
            print("Optimization skipped — wallet already optimal")
        } else {
            print("Optimization completed in \(roundsExecuted) rounds")
        }
    case .inProgress:
        // Full mode runs to completion in one call, so InProgress is
        // not reachable here.
        fatalError("Full mode never returns InProgress")
    }
    // ANCHOR_END: optimize-leaves-full
}

func runOptimizationOneRoundAtATime(sdk: BreezSdk) async throws {
    // ANCHOR: optimize-leaves-single-round
    var roundsExecuted: UInt32 = 0
    loop: while true {
        let outcome = try await sdk.optimizeLeaves(
            request: OptimizeLeavesRequest(mode: .singleRound)
        ).outcome

        switch outcome {
        case .inProgress:
            roundsExecuted += 1
            print("Executed round \(roundsExecuted)")
        case .completed(let thisRound):
            roundsExecuted += thisRound
            if roundsExecuted == 0 {
                print("Optimization skipped — wallet already optimal")
            } else {
                print("Optimization done after \(roundsExecuted) rounds")
            }
            break loop
        }
    }
    // ANCHOR_END: optimize-leaves-single-round
}

func handleAutoOptimizationEvent(event: AutoOptimizationEvent) {
    // ANCHOR: auto-optimization-events
    switch event {
    case .started(let totalRounds):
        print("Auto-optimization started with \(totalRounds) rounds")
    case .roundCompleted(let currentRound, let totalRounds):
        print("Auto-optimization round \(currentRound) of \(totalRounds) completed")
    case .completed:
        print("Auto-optimization completed successfully")
    case .cancelled:
        print("Auto-optimization was cancelled")
    case .failed(let error):
        print("Auto-optimization failed: \(error)")
    case .skipped:
        print("Auto-optimization was skipped because leaves are already optimal")
    }
    // ANCHOR_END: auto-optimization-events
}

File: docs/breez-sdk/snippets/kotlin_mpp_lib/shared/src/commonMain/kotlin/com/example/kotlinmpplib/Optimize.kt
package com.example.kotlinmpplib

import breez_sdk_spark.*

class Optimize {
    suspend fun runFullOptimization(sdk: BreezSdk) {
        // ANCHOR: optimize-leaves-full
        val outcome = sdk.optimizeLeaves(OptimizeLeavesRequest(mode = OptimizationMode.FULL)).outcome

        when (outcome) {
            is OptimizationOutcome.Completed -> {
                if (outcome.roundsExecuted == 0u) {
                    // Log.v("Breez", "Optimization skipped — wallet already optimal")
                } else {
                    // Log.v("Breez", "Optimization completed in ${outcome.roundsExecuted} rounds")
                }
            }
            is OptimizationOutcome.InProgress -> {
                // Full mode runs to completion in one call, so InProgress is
                // not reachable here.
                throw IllegalStateException("Full mode never returns InProgress")
            }
        }
        // ANCHOR_END: optimize-leaves-full
    }

    suspend fun runOptimizationOneRoundAtATime(sdk: BreezSdk) {
        // ANCHOR: optimize-leaves-single-round
        var roundsExecuted: UInt = 0u
        while (true) {
            val outcome = sdk.optimizeLeaves(OptimizeLeavesRequest(mode = OptimizationMode.SINGLE_ROUND)).outcome
            when (outcome) {
                is OptimizationOutcome.InProgress -> {
                    roundsExecuted += 1u
                    // Log.v("Breez", "Executed round $roundsExecuted")
                }
                is OptimizationOutcome.Completed -> {
                    roundsExecuted += outcome.roundsExecuted
                    if (roundsExecuted == 0u) {
                        // Log.v("Breez", "Optimization skipped — wallet already optimal")
                    } else {
                        // Log.v("Breez", "Optimization done after $roundsExecuted rounds")
                    }
                    break
                }
            }
        }
        // ANCHOR_END: optimize-leaves-single-round
    }

    fun handleAutoOptimizationEvent(optimizationEvent: AutoOptimizationEvent) {
        // ANCHOR: auto-optimization-events
        when (optimizationEvent) {
            is AutoOptimizationEvent.Started -> {
                // Log.v("Breez", "Auto-optimization started with ${optimizationEvent.totalRounds} rounds")
            }
            is AutoOptimizationEvent.RoundCompleted -> {
                // Log.v("Breez", "Auto-optimization round ${optimizationEvent.currentRound} of ${optimizationEvent.totalRounds} completed")
            }
            is AutoOptimizationEvent.Completed -> {
                // Log.v("Breez", "Auto-optimization completed successfully")
            }
            is AutoOptimizationEvent.Cancelled -> {
                // Log.v("Breez", "Auto-optimization was cancelled")
            }
            is AutoOptimizationEvent.Failed -> {
                // Log.v("Breez", "Auto-optimization failed: ${optimizationEvent.error}")
            }
            is AutoOptimizationEvent.Skipped -> {
                // Log.v("Breez", "Auto-optimization was skipped because leaves are already optimal")
            }
        }
        // ANCHOR_END: auto-optimization-events
    }
}
